File: packages/python/elizaos/bootstrap/providers/character.py
from __future__ import annotations

from typing import TYPE_CHECKING

from elizaos.types import Provider, ProviderResult

if TYPE_CHECKING:
    from elizaos.types import IAgentRuntime, Memory, State


async def get_character_context(
    runtime: IAgentRuntime,
    message: Memory,
    state: State | None = None,
) -> ProviderResult:
    character = runtime.character

    sections: list[str] = []

    sections.append(f"# Agent: {character.name}")

    if character.bio:
        bio_text = character.bio if isinstance(character.bio, str) else "\n".join(character.bio)
        sections.append(f"\n## Bio\n{bio_text}")

    if character.adjectives:
        adjectives = (
            character.adjectives
            if isinstance(character.adjectives, list)
            else [character.adjectives]
        )
        sections.append(f"\n## Personality Traits\n{', '.join(adjectives)}")

    # lore is optional and may not exist on all Character instances
    lore = getattr(character, "lore", None)
    if lore:
        lore_text = lore if isinstance(lore, str) else "\n".join(lore)
        sections.append(f"\n## Background\n{lore_text}")

    if character.topics:
        topics = character.topics if isinstance(character.topics, list) else [character.topics]
        sections.append(f"\n## Knowledge Areas\n{', '.join(topics)}")

    if character.style:
        style_sections: list[str] = []
        if character.style.all:
            all_style = (
                character.style.all
                if isinstance(character.style.all, list)
                else [character.style.all]
            )
            style_sections.append(f"General: {', '.join(all_style)}")
        if character.style.chat:
            chat_style = (
                character.style.chat
                if isinstance(character.style.chat, list)
                else [character.style.chat]
            )
            style_sections.append(f"Chat: {', '.join(chat_style)}")
        if character.style.post:
            post_style = (
                character.style.post
                if isinstance(character.style.post, list)
                else [character.style.post]
            )
            style_sections.append(f"Posts: {', '.join(post_style)}")
        if style_sections:
            sections.append("\n## Communication Style\n" + "\n".join(style_sections))

    context_text = "\n".join(sections)

    return ProviderResult(
        text=context_text,
        values={
            "agentName": character.name,
            "hasCharacter": True,
        },
        data={
            "name": character.name,
            "bio": character.bio,
            "adjectives": character.adjectives,
            "topics": character.topics,
        },
    )


character_provider = Provider(
    name="CHARACTER",
    description="Provides the agent's character definition and personality information",
    get=get_character_context,
    dynamic=False,
)

File: packages/rust/src/bootstrap/providers/character.rs
//! CHARACTER provider implementation.

use async_trait::async_trait;

use crate::error::PluginResult;
use crate::runtime::IAgentRuntime;
use crate::types::{Memory, ProviderResult, State};

use super::Provider;

/// Provider for character information.
pub struct CharacterProvider;

#[cfg_attr(not(target_arch = "wasm32"), async_trait)]
#[cfg_attr(target_arch = "wasm32", async_trait(?Send))]
impl Provider for CharacterProvider {
    fn name(&self) -> &'static str {
        "CHARACTER"
    }

    fn description(&self) -> &'static str {
        "Provides the agent's character definition and personality information"
    }

    fn is_dynamic(&self) -> bool {
        false
    }

    async fn get(
        &self,
        runtime: &dyn IAgentRuntime,
        _message: &Memory,
        _state: Option<&State>,
    ) -> PluginResult<ProviderResult> {
        let character = runtime.character();
        let mut sections = Vec::new();

        // Name section
        sections.push(format!("# Agent: {}", character.name));

        // Bio section
        if !character.bio.is_empty() {
            sections.push(format!("\n## Bio\n{}", character.bio));
        }

        // Personality/Adjectives section
        if !character.adjectives.is_empty() {
            sections.push(format!(
                "\n## Personality Traits\n{}",
                character.adjectives.join(", ")
            ));
        }

        // Lore/Background section
        if !character.lore.is_empty() {
            sections.push(format!("\n## Background\n{}", character.lore));
        }

        // Topics/Knowledge areas section
        if !character.topics.is_empty() {
            sections.push(format!(
                "\n## Knowledge Areas\n{}",
                character.topics.join(", ")
            ));
        }

        // Style section
        let mut style_parts = Vec::new();
        if !character.style.all.is_empty() {
            style_parts.push(format!("General: {}", character.style.all.join(", ")));
        }
        if !character.style.chat.is_empty() {
            style_parts.push(format!("Chat: {}", character.style.chat.join(", ")));
        }
        if !character.style.post.is_empty() {
            style_parts.push(format!("Posts: {}", character.style.post.join(", ")));
        }
        if !style_parts.is_empty() {
            sections.push(format!("\n## Communication Style\n{}", style_parts.join("\n")));
        }

        let context_text = sections.join("\n");

        Ok(ProviderResult::new(context_text)
            .with_value("agentName", character.name.clone())
            .with_value("hasCharacter", true)
            .with_data("name", character.name.clone())
            .with_data("bio", character.bio.clone()))
    }
}

